### PR TITLE
feat(composer)! avoid holding private key in env var

### LINKED ISSUE #993: Replace `ASTRIA_COMPOSER_PRIVATE_KEY` env var with `ASTRIA_COMPOSER_PRIVATE_KEY_FILE`
We should avoid holding a private key in a configuration env var.  Instead we'll have an env var provide a path to a file which holds the private key.

### PR DESCRIPTION
**Pull request by Fraser999**
## Summary
This PR changes the configuration env vars for composer to not hold the contents of the private key for signing sequencer transactions.

## Background
It's undesirable to have the contents of a private key in an env var.

## Changes
Changed the existing env var to hold a path to a hex-encoded private key file.

## Testing
Modified existing tests, and ran smoke test using local build of composer.

## Breaking Changelist
- Removed `ASTRIA_COMPOSER_PRIVATE_KEY`
- Added `ASTRIA_COMPOSER_PRIVATE_KEY_FILE`

## Related Issues
Closes #993.

### CODE REVIEW

File: crates/astria-composer/local.env.example [hunk 1]
```diff
@@ -35,9 +35,9 @@ ASTRIA_COMPOSER_SEQUENCER_CHAIN_ID="astria-dev-1"
 # names are sha256 hashed and used as the `rollup_id` in `SequenceAction`s
 ASTRIA_COMPOSER_ROLLUPS="astriachain::ws://127.0.0.1:8545"
 
-# Private key for the sequencer account used for signing transactions
-# Must be a hex-encoded 32-byte array (64-character hex string)
-ASTRIA_COMPOSER_PRIVATE_KEY="2bd806c97f0e00af1a1fc3328fa763a9269723c8db8fac4f93af71db186d6e90"
+# The path to the file storing the private key for the sequencer account used for signing
+# transactions.  The file should be a hex-encoded Ed25519 secret key.
```
Comment (SuperFluffy) [inline crates/astria-composer/local.env.example:39]:
```suggestion
# transactions. The file should be contain a hex-encoded Ed25519 secret key.
```

File: crates/astria-composer/src/executor/builder.rs [hunk 2]
```diff
@@ -49,12 +47,17 @@ impl Builder {
...
+        let private_key_bytes: [u8; 32] = hex::decode(private_key_hex.trim())
+            .wrap_err_with(|| {
+                format!("failed to hex-decode private key bytes in `{private_key_file}`")
+            })?
             .try_into()
-            .map_err(|_| eyre!("invalid private key length; must be 32 bytes"))?;
+            .map_err(|_| {
+                eyre!("invalid private key length in `{private_key_file}`; must be 32 bytes")
+            })?;
         let sequencer_key = SigningKey::from(private_key_bytes);
```
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/builder.rs:60]: Maybe tuck this in a function so we can have single `wrap_err_with`?

```rust
fn read_signing_key_from_file<P: AsRef<Path>>(path: P) -> eyre::Result<SingingKey> {
    todo!()
}

read_signing_key_from_file(&private_key_file)
    .wrap_err_with(|| format!("failed reading signing key from file at path `{private_key_file}`)?
```

**Review by SuperFluffy — COMMENTED** [review]
Agree with the changes.

Can we also put the `SigningKey` behind a https://docs.rs/secrecy/0.8.0/secrecy/struct.Secret.html so that we don't accidentally leak it if someone decides to debug-print executor or the submitfut? Right now `SigningKey` implements `Debug` and happily leaks its contents.

File: crates/astria-composer/src/executor/builder.rs (hunk 2 above)
Comment (Fraser999) [reply crates/astria-composer/src/executor/builder.rs]: Done in [840706c](https://github.com/astriaorg/astria/pull/1074/commits/840706c736bcaeb33cb74baf594fa6bc3429ec8b).

**Comment by Fraser999** [comment]
@SuperFluffy - as agreed, I put the signing key in a newtype in [3515ac0](https://github.com/astriaorg/astria/pull/1074/commits/3515ac04d7a09368782b245a9412db26d3e8edb7) and replaced all usages of the original type.

File: crates/astria-core/src/crypto.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    self,
+    Debug,
+    Formatter,
+};
+
+use ed25519_consensus::{
+    Signature,
+    SigningKey as Ed25519SigningKey,
+    VerificationKey,
+    // VerificationKeyBytes,
```
Comment (SuperFluffy) [inline crates/astria-core/src/crypto.rs:11]: can remove

File: crates/astria-core/src/crypto.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /// Returns the verification key associated with this signing key.
+    #[must_use]
+    pub fn verification_key(&self) -> VerificationKey {
+        self.0.verification_key()
+    }
+}
+
+impl Debug for SigningKey {
```
Comment (SuperFluffy) [inline crates/astria-core/src/crypto.rs:56]: Yeah ok - so we hide the signing key, but still allow for indirectly checking it by exposing the verification key. That's smart.

File: crates/astria-core/src/crypto.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    CryptoRng,
+    RngCore,
+};
+use zeroize::{
+    Zeroize,
+    ZeroizeOnDrop,
+};
+
+#[derive(Clone, Zeroize, ZeroizeOnDrop)]
+pub struct SigningKey(Ed25519SigningKey);
```
Comment (SuperFluffy) [inline crates/astria-core/src/crypto.rs:23]: Small doc comment on top, and also a (non-doc) comment to explain its purpose (I thought it prudent to not mention it in the doc comment for the same reason that you decided to not expose a third-party crate).
```suggestion
/// An Ed25519 signing key.
// *Implementation note*: this is currently a refinement type around
// ed25519_consensus::SigningKey overriding its Debug implementation
// to not accidentally leak it.
pub struct SigningKey(Ed25519SigningKey);
```

File: crates/astria-sequencer-relayer/src/validator.rs [hunk 6]
```diff
@@ -1,36 +1,34 @@
...
 pub(crate) struct Validator {
     /// The tendermint validator account address; defined as
     /// Sha256(verification_key)[..20].
-    #[zeroize(skip)]
     pub(crate) address: account::Id,
 
     /// The ed25519 signing key of this validator.
+    // allow: this entire struct is due to get removed as part of
+    // https://github.com/astriaorg/astria/issues/1010
+    #[allow(dead_code)]
```
Comment (SuperFluffy) [inline crates/astria-sequencer-relayer/src/validator.rs:25]: IMO go ahead and remove this. The fact that `Validator` is pub-crate to being with is an oversight (or maybe due to how the integration tests are written).

Nobody should use `astria-sequencer-relayer` as a direct dependency.

File: crates/astria-sequencer-relayer/src/validator.rs [hunk 7]
```diff
@@ -1,36 +1,34 @@
...
     pub(crate) signing_key: SigningKey,
 
-    #[zeroize(skip)]
     /// The ed25519 verification key of this validator.
+    // allow: this entire struct is due to get removed as part of
+    // https://github.com/astriaorg/astria/issues/1010
+    #[allow(dead_code)]
```
Comment (SuperFluffy) [inline crates/astria-sequencer-relayer/src/validator.rs:31]: same argument as above

**Review by SuperFluffy — APPROVED** [review]

File: crates/astria-core/src/crypto.rs (hunk 3 above)
Comment (Fraser999) [reply crates/astria-core/src/crypto.rs]: Done in [6b05e11](https://github.com/astriaorg/astria/pull/1074/commits/6b05e114d3a3860868252108f2fc7be22636e433).

File: crates/astria-core/src/crypto.rs (hunk 5 above)
Comment (Fraser999) [reply crates/astria-core/src/crypto.rs:26]: Done in [6b05e11](https://github.com/astriaorg/astria/pull/1074/commits/6b05e114d3a3860868252108f2fc7be22636e433).

File: crates/astria-sequencer-relayer/src/validator.rs (hunk 6 above)
Comment (Fraser999) [reply crates/astria-sequencer-relayer/src/validator.rs]: Done in [6b05e11](https://github.com/astriaorg/astria/pull/1074/commits/6b05e114d3a3860868252108f2fc7be22636e433).

File: crates/astria-sequencer-relayer/src/validator.rs (hunk 7 above)
Comment (Fraser999) [reply crates/astria-sequencer-relayer/src/validator.rs]: Done in [6b05e11](https://github.com/astriaorg/astria/pull/1074/commits/6b05e114d3a3860868252108f2fc7be22636e433).

**Pull request merged**